src/main/java/reactor/core/publisher/FluxPeekFuseable.java

```diff
@@ -258,7 +263,7 @@ public T poll() {
 					throw Exceptions.propagate(Operators.onOperatorError(s, e, v));
 				}
 			}
-			if (v == null && sourceMode == SYNC) {
+			if (v == null && (d || sourceMode == SYNC)) {
```

Comment:
Your solution is good anyway, I have tried a similar effect anyway when I worked from the MonoSusbcriber directly (it had to double touch a volatile on poll for checking complete). But as discussed offline, this is not unique behavior to MonoSubscriber, PublishOn, UnicastProcessor and other who use onNext(null) might onComplete just after.